### PR TITLE
[PoC] Slack notification templates

### PR DESCRIPTION
**Pull request by madkinsz**
## Summary

Implements notification message templates for just slack notifications for API discussion.

When fully implemented, closes https://github.com/PrefectHQ/prefect/issues/3531

## Changes

- Adds `template` and `template_vars` to slack notifications

## Importance

Users want to amend notifications to include extra information but the best API for this is not entirely clear. This PoC will help us determine the best way to implement this as a feature.

## Checklist

This PR:

- [ ] adds new tests (if appropriate)
- [ ] adds a change file in the `changes/` directory (if appropriate)
- [ ] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

File: src/prefect/utilities/notifications/notifications.py [hunk 1]
```diff
@@ -125,7 +125,12 @@ def slack_message_formatter(
     tracked_obj: TrackedObjectType,
     state: "prefect.engine.state.State",
     backend_info: bool = True,
+    template: str = None,
```
Comment (madkinsz) [inline src/prefect/utilities/notifications/notifications.py:128]: `None` here so the default value can still be defined in this function but also updated from `slack_notifier` -- not great but preserves the existing location of the default message.

File: src/prefect/utilities/notifications/notifications.py [hunk 2]
```diff
@@ -125,7 +125,12 @@ def slack_message_formatter(
...
+    template_vars: Mapping = None,
 ) -> dict:
+    template = template or f"{tracked_obj} is now in a {state} state"
+    template_vars = template_vars or {}
```
Comment (madkinsz) [inline src/prefect/utilities/notifications/notifications.py:132]: We may want to update `template_vars` with `tracked_obj` and `state` here so there cannot be collisions in the `.format(...)` call

File: src/prefect/utilities/notifications/notifications.py [hunk 3]
```diff
@@ -125,7 +125,12 @@ def slack_message_formatter(
     tracked_obj: TrackedObjectType,
     state: "prefect.engine.state.State",
     backend_info: bool = True,
+    template: str = None,
+    template_vars: Mapping = None,
 ) -> dict:
+    template = template or f"{tracked_obj} is now in a {state} state"
```
Comment (madkinsz) [inline src/prefect/utilities/notifications/notifications.py:131]: Where should we define what names will be provided by default? What if a user wants to add to the template without having to recreate this string?

Comment (madkinsz) [reply src/prefect/utilities/notifications/notifications.py:131]: Note to include prefect context variables

Comment (joshmeek) [reply src/prefect/utilities/notifications/notifications.py:131]: I think that mentioning it in the docstring will suffice. The only things we have available to pass into the format call are those two vars and context

**Comment by madkinsz** [comment]
Closing this as it's only a proof of concept and I'm worried this is discouraging contribution from others.

**Pull request closed without merging**